sprint_17

KatSchreiner opened the pull request.

KatSchreiner committed:
Реализовано контекстное меню
Нужно доработать пункт меню редактирование

KatSchreiner committed:
Реализована фильтрация трекеров
Настройка черной темы

KatSchreiner committed:
Доработана темная тема

KatSchreiner committed:
Реализован экран статистики

KatSchreiner committed:
Восстановила enum TrackerFilter, куда-то пропал

KatSchreiner committed:
Добавлена Аналитика

KatSchreiner committed:
Доработан метод редактирования трекера
Подготовка к ревью

andrew-kuzn-yndx commented on Tracker/CategoryViewController/CategoriesViewController.swift:
```diff
@@ -173,18 +168,24 @@ extension CategoriesViewController: UITableViewDelegate {
+            cell.layer.maskedCorners = [.layerMinXMinYCorner, .layerMaxXMinYCorner, .layerMinXMaxYCorner, .layerMaxXMaxYCorner]
+        } else if numberOfRows > 1 {
             if indexPath.row == 0 {
                 cell.layer.cornerRadius = 16
```
Старайся не работать со свойствами ячейки и вообще класса вот так напрямую из другого класса. 
По хорошему - это сделать в ячейке публичную функцию, например configCell с аргументами на вход, и уже в ней менять нужные свойства. 
Свойства класса нужно всегда стараться закрывать извне, и работать с ними только изнутри

andrew-kuzn-yndx commented on Tracker/CreateTracker/CategoryNameFieldCell.swift:
```diff
@@ -1,10 +1,3 @@
-//
-//  Created by Екатерина Шрайнер on 04.07.2024.
-//
-
```
Заметил, что ты удаляешь шапку swift файлов, где находится информация о разработчике и другая важная информация. Когда ты работаешь на проекте одна, это не важно. Но в будущем эту информацию лучше оставлять, она может пригодиться

KatSchreiner replied:
Хорошо

andrew-kuzn-yndx commented on Tracker/CreateTracker/CreateNewHabitViewController.swift:
```diff
@@ -1,18 +1,11 @@
 import UIKit
 
-final class CreateNewHabitViewController: CreateNewTrackerViewController {
+class CreateNewHabitViewController: CreateNewTrackerViewController {
```
А зачем удалила final?(
Рекомендую добавлять этот модификатор ко всем классам при их создании, а затем, если вдруг понадобится наследоваться от какого-то класса, то компилятор подскажет, что нужно убрать final😊

KatSchreiner replied:
убрала потому что от этого класса наследуется EditTrackerViewController

final class EditTrackerViewController: CreateNewHabitViewController {
}

andrew-kuzn-yndx commented on Tracker/CreateTracker/CreateNewTrackerViewController.swift:
```diff
@@ -19,16 +12,21 @@ class CreateNewTrackerViewController: UIViewController {
     var selectedEmoji: String?
     var selectedColor: UIColor?
     
+    var isEdit = false
```
Не забывай добавлять свойство private, для того, чтобы спрятать детали реализации конкретной части функциональности, когда они используются только внутри области объявления
Стоит проверить весь проект на предмет пропущенных private и final

KatSchreiner replied:
Хорошо

andrew-kuzn-yndx commented on Tracker/CreateTracker/CreateNewTrackerViewController.swift:
```diff
@@ -188,26 +189,40 @@ extension CreateNewTrackerViewController: UICollectionViewDataSource {
             return cell
+            
         case .createNewCategory:
             let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CreateNewCategoryCell.newCategoryIdentifier, for: indexPath) as! CreateNewCategoryCell
```
Тут есть force операция - force cast, она как и force unwrap является нежелательной и надо стараться такого кода избегать. Можно переписать без их использования, переместив в мягкий каст в if: if let a = a as? 
Всегда когда есть возможность написать безопасную конструкцию – отдавай предпочтение ей. Так твоя работа будет выигрышнее смотреться)

andrew-kuzn-yndx commented on Tracker/CreateTracker/CreateNewTrackerViewController.swift:
```diff
@@ -188,26 +189,40 @@ extension CreateNewTrackerViewController: UICollectionViewDataSource {
             return cell
+            
         case .color:
             let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "colorCell", for: indexPath) as! ColorsCollectionViewCell
```
Давай проверим весь проект на наличие force cast)

KatSchreiner replied:
Сделано

andrew-kuzn-yndx commented on Tracker/StatisticView/StatisticViewController.swift:
```diff
@@ -1,26 +1,118 @@
-        view.backgroundColor = .white
+        view.backgroundColor = .yWhite
+        
+        [placeholderStackView, tableView].forEach { view in
```
Тут необходимо прописывать [weak self], чтобы избежать цикла сильных ссылок. 
Можно взять за правило - что всегда, при обращении к self. из замыкания, нужно писать [weak self], чтобы не допустить утечек памяти

KatSchreiner replied:
Готово

andrew-kuzn-yndx commented on Tracker/StatisticView/StatisticViewController.swift:
```diff
@@ -1,26 +1,118 @@
+            tableView.isHidden = true
+        } else {
+            placeholderStackView.isHidden = true
+            tableView.isHidden = false
```
Тут можно упростить:

placeholderStackView.isHidden = !statisticsData.isEmpty
            tableView.isHidden = statisticsData.isEmpty

KatSchreiner replied:
Поправила

KatSchreiner committed:
Ревью
Внесены правки в: CategoriesViewController, CategoryCell.
Исправлено: для работы со свойствами ячейки создала метод configCell.

KatSchreiner committed:
Ревью
Внесены правки в: EditTrackerViewController, CreateNewTrackerViewController, TrackersViewController
Исправлено: переменную isEdit переместила в EditTrackerViewController и добавила ей свойcтво private. Удалила  "createNewTrackerVC.isEdit = true" из TrackersViewController.

KatSchreiner committed:
Ревью
Исправлено: вместо использования force cast (as!), использовала безопасное приведение через guard let.

KatSchreiner committed:
Ревью
Исправлено: прописала  [weak self], чтобы избежать цикла сильных ссылок.

KatSchreiner committed:
Ревью
Внесены правки в: StatisticViewController
Исправлено: использовала упрощенный вариант видимости placeholder и tableView

KatSchreiner committed:
Удалила папку шрифтов т.к. она не используется

KatSchreiner committed:
Ревью
Исправлено: При смене дней недели, показывается плейсхолдер "Что будем отслеживать"
Исправлено: После использования поиска - заглушка больше не накладывается на список трекеров

KatSchreiner committed:
Ревью

Исправлено: добавила отображение показателя статистики «Трекеров завершено».
Исправлено: счётчик статистики обновляется при взаимодействии пользователя с треком.

KatSchreiner committed:
Ревью

Решена проблема отсутствия кнопки фильтров после создания трекера на будущую и прошлую дату
